File: internal/detect/glob_test.go
package detect

import (
	"reflect"
	"testing"
)

func TestGlobToRegexp(t *testing.T) {
	cases := []struct {
		pattern string
		match   []string
		reject  []string
	}{
		{"packages/*", []string{"packages/a", "packages/a-b"}, []string{"packages", "packages/a/b", "packagesx/a"}},
		{"a?c", []string{"abc", "a-c"}, []string{"ac", "abbc", "a/c"}},
		// The zero-segment case is the one everyone gets wrong: a/**/b must
		// match a/b as well as a/z/b.
		{"a/**/b", []string{"a/b", "a/z/b", "a/z/y/b"}, []string{"a/zb", "ab"}},
		{"packages/**", []string{"packages/a", "packages/a/b"}, []string{"packages"}},
		{"**x", []string{"x", "ax", "a/bx"}, []string{"xy"}},
		{"p+kg/*", []string{"p+kg/a"}, []string{"ppkg/a", "pkg/a"}},
		{"café/*", []string{"café/a"}, []string{"cafe/a"}},
		// A character class is matched literally, bracket and all.
		{"a[b]c", []string{"a[b]c"}, []string{"abc", "ab"}},
		{"my dir/*", []string{"my dir/a"}, []string{"mydir/a"}},
	}
	for _, c := range cases {
		re := GlobToRegexp(c.pattern)
		for _, s := range c.match {
			if !re.MatchString(s) {
				t.Errorf("GlobToRegexp(%q) = %v must match %q", c.pattern, re, s)
			}
		}
		for _, s := range c.reject {
			if re.MatchString(s) {
				t.Errorf("GlobToRegexp(%q) = %v must not match %q", c.pattern, re, s)
			}
		}
	}
}

func collectNotes() (func(string), *[]string) {
	var notes []string
	return func(msg string) { notes = append(notes, msg) }, &notes
}

func TestBuildPatternsResolvesAgainstTheRoot(t *testing.T) {
	note, _ := collectNotes()
	ps := BuildPatterns("/", []string{"packages/*"}, note)
	if got := ps.Expand([]string{"/packages/a", "/packages/a/b", "/other"}); !reflect.DeepEqual(got, []string{"/packages/a"}) {
		t.Errorf("Expand = %v", got)
	}

	ps = BuildPatterns("/tools", []string{"pkgs/*"}, note)
	if got := ps.Expand([]string{"/tools/pkgs/a", "/pkgs/a"}); !reflect.DeepEqual(got, []string{"/tools/pkgs/a"}) {
		t.Errorf("Expand = %v", got)
	}
}

func TestBuildPatternsNegation(t *testing.T) {
	note, _ := collectNotes()
	ps := BuildPatterns("/", []string{"packages/*", "!packages/legacy"}, note)
	got := ps.Expand([]string{"/packages/a", "/packages/legacy", "/packages/z"})
	if want := []string{"/packages/a", "/packages/z"}; !reflect.DeepEqual(got, want) {
		t.Errorf("Expand = %v, want %v", got, want)
	}
}

func TestBuildPatternsCleansAndSkips(t *testing.T) {
	note, notes := collectNotes()
	ps := BuildPatterns("/", []string{"./pkgs/*", "trail//", "", "!"}, note)
	got := ps.Expand([]string{"/pkgs/a", "/trail", "/other"})
	if want := []string{"/pkgs/a", "/trail"}; !reflect.DeepEqual(got, want) {
		t.Errorf("Expand = %v, want %v", got, want)
	}
	if len(*notes) != 0 {
		t.Errorf("unexpected notes %v", *notes)
	}
}

func TestBuildPatternsNotesCharacterClasses(t *testing.T) {
	note, notes := collectNotes()
	BuildPatterns("/", []string{"pkg[ab]/*"}, note)
	want := "workspace pattern 'pkg[ab]/*' uses a character class, which is matched literally"
	if len(*notes) != 1 || (*notes)[0] != want {
		t.Errorf("notes = %v, want [%q]", *notes, want)
	}
}

func TestExpandWithNoPositivePatternsMatchesNothing(t *testing.T) {
	note, _ := collectNotes()
	ps := BuildPatterns("/", []string{"!packages/legacy"}, note)
	if got := ps.Expand([]string{"/packages/a", "/packages/legacy"}); len(got) != 0 {
		// A negation with no positive patterns has nothing to subtract from.
		t.Errorf("Expand = %v, want empty", got)
	}
}

File: tests/Unit/GlobTest.php
<?php

declare(strict_types=1);

use Automattic\DependabotDirectories\Glob\GlobTranslator;
use Automattic\DependabotDirectories\Glob\PatternSet;

it('translates a glob into an anchored matcher', function (string $pattern, array $match, array $reject) {
    $compiled = GlobTranslator::toRegexp($pattern);

    foreach ($match as $subject) {
        expect(GlobTranslator::matches($compiled, $subject))
            ->toBeTrue("{$pattern} must match {$subject}");
    }
    foreach ($reject as $subject) {
        expect(GlobTranslator::matches($compiled, $subject))
            ->toBeFalse("{$pattern} must not match {$subject}");
    }
})->with([
    'single star stays inside one segment' => ['packages/*', ['packages/a', 'packages/a-b'], ['packages', 'packages/a/b', 'packagesx/a']],
    'question mark is one character' => ['a?c', ['abc', 'a-c'], ['ac', 'abbc', 'a/c']],
    // The zero-segment case is the one everyone gets wrong: a/**/b must match
    // a/b as well as a/z/b.
    'globstar covers zero segments' => ['a/**/b', ['a/b', 'a/z/b', 'a/z/y/b'], ['a/zb', 'ab']],
    'trailing globstar crosses slashes' => ['packages/**', ['packages/a', 'packages/a/b'], ['packages']],
    'leading globstar' => ['**x', ['x', 'ax', 'a/bx'], ['xy']],
    'regex metacharacters are literal' => ['p+kg/*', ['p+kg/a'], ['ppkg/a', 'pkg/a']],
    'unicode survives' => ['café/*', ['café/a'], ['cafe/a']],
    // A character class is matched literally, bracket and all.
    'character class is literal' => ['a[b]c', ['a[b]c'], ['abc', 'ab']],
    'spaces survive' => ['my dir/*', ['my dir/a'], ['mydir/a']],
]);

it('resolves patterns against the workspace root', function () {
    $set = PatternSet::build('/', ['packages/*'], fn () => null);
    expect($set->expand(['/packages/a', '/packages/a/b', '/other']))->toBe(['/packages/a']);

    $set = PatternSet::build('/tools', ['pkgs/*'], fn () => null);
    expect($set->expand(['/tools/pkgs/a', '/pkgs/a']))->toBe(['/tools/pkgs/a']);
});

it('subtracts negated patterns', function () {
    $set = PatternSet::build('/', ['packages/*', '!packages/legacy'], fn () => null);

    expect($set->expand(['/packages/a', '/packages/legacy', '/packages/z']))
        ->toBe(['/packages/a', '/packages/z']);
});

it('cleans and skips empty patterns without complaining', function () {
    $notes = [];
    $set = PatternSet::build('/', ['./pkgs/*', 'trail//', '', '!'], function (string $n) use (&$notes) {
        $notes[] = $n;
    });

    expect($set->expand(['/pkgs/a', '/trail', '/other']))->toBe(['/pkgs/a', '/trail']);
    expect($notes)->toBe([]);
});

it('notes a character class rather than pretending to support it', function () {
    $notes = [];
    PatternSet::build('/', ['pkg[ab]/*'], function (string $n) use (&$notes) {
        $notes[] = $n;
    });

    expect($notes)->toBe(["workspace pattern 'pkg[ab]/*' uses a character class, which is matched literally"]);
});

it('matches nothing when there are no positive patterns', function () {
    // A negation alone has nothing to subtract from.
    $set = PatternSet::build('/', ['!packages/legacy'], fn () => null);

    expect($set->expand(['/packages/a', '/packages/legacy']))->toBe([]);
});

it('fails loudly rather than silently reporting no match on a PCRE error', function () {
    // preg_match returns false on a blown backtrack limit, and false is
    // falsy, so letting it flow on would read as "no match" and quietly drop
    // a directory from coverage. The limit is forced down here because
    // realistic path depths never reach the default.
    $limit = ini_get('pcre.backtrack_limit');
    ini_set('pcre.backtrack_limit', '1');

    try {
        $compiled = GlobTranslator::toRegexp('a/**/b');
        $subject = 'a/'.str_repeat('deep/', 50).'b';

        expect(fn () => GlobTranslator::matches($compiled, $subject))
            ->toThrow(RuntimeException::class);
    } finally {
        ini_set('pcre.backtrack_limit', false === $limit ? '1000000' : $limit);
    }
});
